### ISSUE TITLE
getJsonToken assumes underlying string is null terminated

### ISSUE THREAD

**Issue by Brotcrunsher** [bitcoin/bitcoin#28260]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

When a function requires a `const char *end`, one might assume that the string doesn't need to be null terminated. `getJsonToken` requires this parameter:
```
enum jtokentype getJsonToken(std::string& tokenVal, unsigned int& consumed,
                            const char *raw, const char *end)
```

However, the function does assume that the underlying memory block is null-terminated as is visible here:
```
        if (!strncmp(raw, "null", 4)) {
            //...
        } else if (!strncmp(raw, "true", 4)) {
            //...
        } else if (!strncmp(raw, "false", 5)) {
            //...
```
and here:
```
if ((*firstDigit == '0') && json_isdigit(firstDigit[1]))
```
and possibly others.

If the string isn't null terminated, then we run risk of reading uninitialized memory, which could lead to a crash.

In practice this currently does not seem to lead to any issues, as all usages that I could find used underlying null-terminated blocks. It is however an uncomfortable pitfall for future development.

### Expected behaviour

The function should either make sure to never access anything beyond or at the end pointer, or should document this requirement properly.

### Steps to reproduce

```
char innocentChar = 'n';
getJsonToken(/*...*/, &innocentChar, (&innocentChar) + 1);
```

### How did you obtain Bitcoin Core

Other

### What version of Bitcoin Core are you using?

master@3654d84

### Operating system and version

Common code issue, os irrelevant

**Comment by daniel-btc-nym** [comment]
I'm working on it.  Plan is to harden the function to not read at end and after. I have some initial small draft changes, and want to add testing. Maybe, we can even highlight the issue via additional fuzz+msan tests.

**Comment by daniel-btc-nym** [comment]
The bug here isn't actually that bad: 

- It uses strncmp(..., 4) and strncmp(..., 5) respectively: so, in worst case it will read 3 or 5 bytes after end.
- Similar for the number parsing.

In any case, I have a change that fixes both  (adding guards for reading past end, and making the code pretty / easy to read again). I have unit tests for the keywords (true, false, null) but want to add some for the numbers.

Haven't made any progress on MSAN. But if we would poison data around begin and end, we could write some exhaustive tests for small inputs. If we restrict to (printable characters, or these mentioned in the code), we could then be pretty certain the new code is correct. 

Touching code is always a risk, so I am trying to minimize it.

**Comment by daniel-btc-nym** [comment]
Yeah, I ran some valgrind tests before and after my changes and it clearly showed the "uninitialized" accesses when we copy the data first into a tight vector as you suggested.

I added more test coverage and think this should be enough for this issue.

What do you think about  [the changes](https://github.com/bitcoin/bitcoin/compare/master...daniel-btc-nym:bitcoin:master)

If they look OK, I can merge them into a single commit  --  (I suppose by forking again?)  --  and then making a pull-request?

**Comment by daniel-btc-nym** [comment]
> I like bugfixes the most, when they are touching only a single line of code. But if more changes are required, then that is fine, too.

My changes have a few orthogonal, but related fixes and test additions. We could make different pull-requests for them. Let me know what you prefer.

For example:
1. Check strncmp will not go out of bound. This is lines 99-115 in univalue_read.cpp ; unit tests that call `no_reading_beyond_end_test` 
1. Change parsing of numbers, where there were two places that might read past the end. This is lines 126-143 in univalue_read.cpp ; unit test that call: `parsing_numbers_beyond_end_test`.
1. Added exhaustive_short_string_test unit test, which was previously causing errors when run in valgrind; now runs cleanly.
1. Added a couple more unit tests such as parsing the empty buffer.
1. Fixed: `fail18.json` which is a test where the nesting is too deep.
1. Added `pass5.json` which is a test where the nesting is just shallow enough to pass.

> It may be good to update the fuzz target to catch the issue here: `src/test/fuzz/parse_univalue.cpp`

I added it in, and ran the fuzzer using HONGGFUZZ. Since this isn't a "runaway" overflow, we don't get any crashes. When running the entire fuzzer under valgrind, we seem to get all sorts of (seemingly unrelated) problems.

I thus added a mini-exhausted test that runs quickly and cleanly under valgrind now, but didn't before.

A few more questions:

- Are we adhering to any line lengths conventions?
- Are we adhering to conventions for variable names and function names?
- My changes *do* change the semantics for the corner-cases where we were reading past the end of the buffer before: For example, previously:  `char buf[] = "true"; getJsonToken(tokenVal, consumed, buf, buf + 1);` would parse as `JTOK_KW_TRUE`  and now returns `JTOK_ERR`. There are also the opposite: previously, `char buf[] = "01"; getJsonToken(tokenVal, consumed, buf, buf + 1);` would error since we do not allow leading 0s, but now, this would parse as "0".
- Do we have some place where we could dump outputs from valgrind that I could link to in the commit messages or pull-request explanation? I would like to show before / after? Or basically, anywhere that creates durable data that can easily be linked to?
- What do you suggest my next steps would be?  I am fairly happy with the [code](https://github.com/bitcoin/bitcoin/compare/master...daniel-btc-nym:bitcoin:master).
